cloudnetwork/v1/types.go

```diff
@@ -0,0 +1,103 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// CloudPrivateIPConfig allows the user to assign private IP addresses
+// to the primary NIC associated with cloud VMs. This is done by
+// specifying a list Kubernetes nodes and IPs requested for those nodes.
+// This CRD is intended on being used by the network plugin which manages
+// the cluster network.
+// Note: this CRD is managed using a singleton cluster-wide CR, which is
+// called: `cluster`. This is enforced by the CRD's unique name in
+// combination with the fact that the CRD is cluster-scoped.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:path=cloudprivateipconfig,scope=Cluster
+type CloudPrivateIPConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// spec is the definition of the desired private IP request(s).
+	// +kubebuilder:validation:Optional
+	// +optional
+	Spec CloudPrivateIPConfigSpec `json:"spec" protobuf:"bytes,2,opt,name=spec,casttype=CloudPrivateIPConfigSpec"`
+	// status is the observed status of the desired private IP request(s). Read-only.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Status CloudPrivateIPConfigStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status,casttype=CloudPrivateIPConfigStatus"`
+}
+
+// CloudPrivateIPConfigSpec specifies the requested list of private IPs and their corresponding node assignment.
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigSpec struct {
+	// ipConfigs is a unique set of tuples (node, ip). A node can have several
+	// different IPs defined by creating a separate tuple for each.
+	// The order is not strict.
```

Comment:
perhaps: "order is not semantically important"

You need to enforce ordering of the items to avoid unnecessary changes when computing new sets.  For instance [a, b] may not semantically differ from [b,a], but you want to always store it as [a,b] to avoid unnecessary object updates.